crates/vm/levm/tests/tests.rs

```diff
@@ -3362,3 +3363,146 @@ fn create_on_create() {
     vm.execute();
     assert_eq!(vm.db.accounts.len(), 4);
 }
+
+#[test]
+fn callvalue_op() {
+    let address_that_has_the_code = Address::from_low_u64_be(0x42);
+    let value = U256::from(0x1234);
+
+    let operations = [Operation::Callvalue, Operation::Stop];
+
+    let tx_env = TxEnv {
+        transact_to: TransactTo::Call(address_that_has_the_code),
+        value,
+        ..Default::default()
+    };
+
+    let block_env = BlockEnv::default();
+
+    let mut db = Db::default();
+
+    db.add_account(
+        address_that_has_the_code,
+        Account::default()
+            .with_bytecode(ops_to_bytecde(&operations))
+            .with_balance(U256::MAX),
+    );
+
+    let mut vm = VM::new(tx_env, block_env, db);
+
+    vm.execute();
+
+    assert_eq!(vm.current_call_frame_mut().stack.pop().unwrap(), value);
+    assert_eq!(vm.env.consumed_gas, TX_BASE_COST + gas_cost::CALLVALUE);
+}
+
+#[test]
+fn codesize_op() {
+    let address_that_has_the_code = Address::from_low_u64_be(0x42);
+
+    let operations = [Operation::Codesize, Operation::Stop];
+
+    let tx_env = TxEnv {
+        transact_to: TransactTo::Call(address_that_has_the_code),
+        ..Default::default()
+    };
+
+    let block_env = BlockEnv::default();
+
+    let mut db = Db::default();
+
+    db.add_account(
+        address_that_has_the_code,
+        Account::default()
+            .with_bytecode(ops_to_bytecde(&operations))
+            .with_balance(U256::MAX),
+    );
+
+    let mut vm = VM::new(tx_env, block_env, db);
+
+    vm.execute();
+
+    assert_eq!(
+        vm.current_call_frame_mut().stack.pop().unwrap(),
+        U256::from(2)
+    );
+    assert_eq!(vm.env.consumed_gas, TX_BASE_COST + gas_cost::CODESIZE);
+}
+
+#[test]
+fn gasprice_op() {
+    let address_that_has_the_code = Address::from_low_u64_be(0x42);
+    let operations = [Operation::Gasprice, Operation::Stop];
+
+    let tx_env = TxEnv {
+        transact_to: TransactTo::Call(address_that_has_the_code),
+        gas_price: Some(U256::from(0x9876)),
+        ..Default::default()
+    };
+
+    let block_env = BlockEnv::default();
+
+    let mut db = Db::default();
+
+    db.add_account(
+        address_that_has_the_code,
+        Account::default()
+            .with_bytecode(ops_to_bytecde(&operations))
+            .with_balance(U256::MAX),
+    );
+
+    let mut vm = VM::new(tx_env, block_env, db);
+
+    vm.execute();
+
+    assert_eq!(
+        vm.current_call_frame_mut().stack.pop().unwrap(),
+        U256::from(0x9876)
+    );
+    assert_eq!(vm.env.consumed_gas, TX_BASE_COST + gas_cost::GASPRICE);
+}
+
+#[test]
+fn codecopy_op() {
+    let address_that_has_the_code = Address::from_low_u64_be(0x42);
+    // https://www.evm.codes/playground?unit=Wei&callData=0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF&codeType=Mnemonic&code=%27z1~32~0ywwz2~8~31y%27~wPUSH1%20z%2F%2F%20Example%20y~0wCALLDATACOPYw%5Cn%01wyz~_
```

Comment:
It's not the same example.

Reply:
Whoops you're right

Reply:
Is calldata needed for something in this case?

Reply:
Fixed the url!